AsyncDisplayKit/Layout/ASLayoutValidation.mm

```diff
@@ -0,0 +1,223 @@
+    }
+    
+    // Sublayout layoutable should be wrapped in a ASStaticLayoutSpec
+    if (assertMessage == nil || [layoutable isKindOfClass:stackContainerClass]) {
+      continue;
+    }
+    
+    ASDisplayNodeCAssert(NO, assertMessage);
+  }
+}
+
+@end
+
+
+#pragma mark - ASLayoutableStackValidator
+
+@implementation ASLayoutableStackValidator
+
+#pragma mark <ASLayoutableValidator>
+
+- (void)validateLayout:(ASLayout *)layout
+{
+  id<ASLayoutable> layoutable = layout.layoutableObject;
+  for (ASLayout *sublayout in layout.sublayouts) {
+    id<ASLayoutable> sublayoutLayoutable = sublayout.layoutableObject;
+    
+    NSString *assertMessage = nil;
+    Class stackContainerClass = [ASStackLayoutSpec class];
+    
+    // Check if default values related to ASStackLayoutSpec have changed
+    if (sublayoutLayoutable.spacingBefore != 0) {
+      assertMessage = ASLayoutValidationWrappingAssertMessage(@selector(spacingBefore), sublayoutLayoutable, stackContainerClass);
+    } else if (sublayoutLayoutable.spacingAfter != 0) {
+      assertMessage = ASLayoutValidationWrappingAssertMessage(@selector(spacingAfter), sublayoutLayoutable, stackContainerClass);
+    } else if (sublayoutLayoutable.flexGrow == YES) {
+      assertMessage = ASLayoutValidationWrappingAssertMessage(@selector(flexGrow), sublayoutLayoutable, stackContainerClass);
+    } else if (sublayoutLayoutable.flexShrink == YES) {
+      assertMessage = ASLayoutValidationWrappingAssertMessage(@selector(flexShrink), sublayoutLayoutable, stackContainerClass);
+    } else if (!ASRelativeDimensionEqualToRelativeDimension(sublayoutLayoutable.flexBasis, ASRelativeDimensionUnconstrained) ) {
+      assertMessage = ASLayoutValidationWrappingAssertMessage(@selector(flexBasis), sublayoutLayoutable, stackContainerClass);
+    } else if (sublayoutLayoutable.alignSelf != ASStackLayoutAlignSelfAuto) {
+      assertMessage = ASLayoutValidationWrappingAssertMessage(@selector(alignSelf), sublayoutLayoutable, stackContainerClass);
+    }
+    
+    // Sublayout layoutable should be wrapped in a ASStackLayoutSpec
+    if (assertMessage == nil || [layoutable isKindOfClass:stackContainerClass]) {
+      continue;
+    }
+    
+    ASDisplayNodeCAssert(NO, assertMessage);
+  }
+}
+
+@end
+
+#pragma mark ASLayoutablePreferredSizeValidator
+
+@implementation ASLayoutablePreferredSizeValidator
+
+#pragma mark <ASLayoutableValidator>
+
+- (void)validateLayout:(ASLayout *)layout
+{
+  // TODO: Implement validation that certain node classes need to have a preferredSize set e.g. ASVideoNode
+}
+
+@end
+
+
+#pragma mark - ASLayoutableValidation
+
+@interface ASLayoutableValidation ()
+@property (copy, nonatomic) NSMutableArray<id<ASLayoutableValidator>> *validators;
+@end
+
+@implementation ASLayoutableValidation
+
+#pragma mark Lifecycle
+
+- (instancetype)init
+{
+  self = [super init];
+  if (self) {
+    _validators = [NSMutableArray array];
```

Comment:
Currently we just iterate through it below. Iterating through arrays / vectors should be faster as using set's. NSMutableSet/NSCountedSet would be preferable if we do a lot of "contains" checks as it's based on a hash map. Further information also: [Sets: Unordered Collections of Objects](https://developer.apple.com/library/ios/documentation/Cocoa/Conceptual/Collections/Articles/Sets.html#//apple_ref/doc/uid/20000136-CJBDHAJD)

Please let me know in case I'm missing something.